### PR TITLE
Fix Uts behavior for qnx7

### PR DESCRIPTION
**Pull request by SNiukalov**
This PR is **ready** for review.  
This PR includes implementation for  proposal [Porting SDL to QNX700 and Automotive Grade Linux (AGL) x64 platforms](https://github.com/LuxoftAKutsan/sdl_evolution/blob/port_sdl_on_qnx700_and_agl/proposals/nnnn-sdl_on_qnx700_agl.md)

### Summary
Unit tests have been fixed for: 

- hmi_message_handler_impl_test

Commit "Fixed size stack of thread":https://github.com/Ford-Luxoft/sdl_core/pull/34/commits/74663f321d40909deb3b4cea9d660e4cab27b183 
_corrected stack size of a thread, which is then used as params  for the pthread_attr_setstacksize in Thread::start function.
We are using more than 4K memory for running a test, but set less than 1K.
For the correct work 6K is sufficient, but I set 8K because of precaution._

Commit "Fixed logic of work using boost conditional variable":https://github.com/Ford-Luxoft/sdl_core/pull/34/commits/052292834a23b84d938f226e6884c483b72ce2ec & https://github.com/Ford-Luxoft/sdl_core/pull/34/commits/511a9ffa901613d55d8e31c805d80669274d0ac1
_implement it:_ 
_Note that the same mutex is locked before the shared data is updated, but that  mutex does not  have to be locked across the call to notify_one._:https://www.boost.org/doc/libs/1_66_0/doc/html/thread/synchronization.html#thread.synchronization.condvar_ref 

Commit "Fixed behavior of destruction for thread and delegate thread.":https://github.com/Ford-Luxoft/sdl_core/pull/34/commits/a8765c36b6b7485a9c44d10e99ad828e10c812c9
function  threads::DeleteThread(thread_) deletes thread object which is used inside "thread_delegate_ " and "join()" in ~Thread()

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/utils/src/threads/thread_posix.cc [hunk 1]
```diff
@@ -158,83 +162,90 @@ bool Thread::IsCurrentThread() const {
 
 bool Thread::start(const ThreadOptions& options) {
   LOG4CXX_AUTO_TRACE(logger_);
+  pthread_attr_t attributes;
+  int pthread_result = -1;
+  {
+    sync_primitives::AutoLock auto_lock(run_lock_);
+    // 1 - state_lock locked
```
Comment (mked-luxoft) [inline src/components/utils/src/threads/thread_posix.cc]: @SNiukalov as far as i understand, we now acquire `run_lock_` instead of `state_lock_`. If it is true, this comment section is rather deceiving. I think i'd rather remove it.

Comment (SNiukalov) [reply src/components/utils/src/threads/thread_posix.cc]: Thanks fixed in: https://github.com/Ford-Luxoft/sdl_core/pull/34/commits/5b98c28c89c072e1baafc5ea6238058ae73c1447

**Review by mked-luxoft — APPROVED** [review]

**Review by LuxoftAKutsan — APPROVED** [review]
Please insert information from PR description into commits.

File: src/components/include/utils/threads/message_loop_thread.h [hunk 2]
```diff
@@ -147,9 +147,8 @@ MessageLoopThread<Q>::MessageLoopThread(const std::string& name,
 
 template <class Q>
 MessageLoopThread<Q>::~MessageLoopThread() {
-  Shutdown();
```
Comment (AByzhynar) [inline src/components/include/utils/threads/message_loop_thread.h:150]: @SNiukalov Why did you remove it?

File: src/components/include/utils/threads/message_loop_thread.h [hunk 3]
```diff
@@ -147,9 +147,8 @@ MessageLoopThread<Q>::MessageLoopThread(const std::string& name,
...
-  delete thread_delegate_;
   threads::DeleteThread(thread_);
```
Comment (AByzhynar) [inline src/components/include/utils/threads/message_loop_thread.h:150]: @SNiukalov `thread_` contains `delegate` . So why did you delete first containing object and only after you delete part of it (its delegate)?

File: src/components/utils/src/threads/thread_posix.cc [hunk 4]
```diff
@@ -86,12 +86,16 @@ void* Thread::threadFunc(void* arg) {
 
   pthread_cleanup_push(&cleanup, thread);
 
-  thread->state_lock_.Acquire();
+  {
+    sync_primitives::AutoLock auto_lock(thread->run_lock_);
```
Comment (AByzhynar) [inline src/components/utils/src/threads/thread_posix.cc:90]: @SNiukalov Lock for acquiring another lock?

**Review by AByzhynar — CHANGES_REQUESTED** [review]

**Review by AKalinich-Luxoft — CHANGES_REQUESTED** [review]
Please address Andrew's comments

**Comment by vkushnirenko-luxoft** [comment]
AGL_build

**Comment by vkushnirenko-luxoft** [comment]
Linux_build

**Comment by vkushnirenko-luxoft** [comment]
QNX7_build

**Comment by vkushnirenko-luxoft** [comment]
Linux_build

File: src/components/include/utils/threads/message_loop_thread.h (hunk 2 above)
Comment (SNiukalov) [reply src/components/include/utils/threads/message_loop_thread.h:150]: Shutdown () was removed because it is redundant. Destructor Thread will make this.

File: src/components/include/utils/threads/message_loop_thread.h (hunk 3 above)
Comment (SNiukalov) [reply src/components/include/utils/threads/message_loop_thread.h:150]: We used aggregation in Thread class and we dont have contorol over delegate life cycle, so logically need first to delete container and then delete delegate. we are calling join() -> stop() Inside destuctor of Thread we are using delegate_->exitThreadMain(); Inside the stop()

File: src/components/utils/src/threads/thread_posix.cc (hunk 4 above)
Comment (SNiukalov) [reply src/components/utils/src/threads/thread_posix.cc:90]: This is not a lock for acquiring another lock. This code part ensures correct notification sequence between Thread::start and threadFunc. In Thread::start function after thread is created, we call Wait(), but we cannot guarantee order of precedence of the following calls : "Wait() in Thread::start" or "Broadcast in threadFunc". sync_primitives::AutoLock auto_lock(thread->run_lock_); - is used to guarantee call to "Wait() in Thread::start" before Broadcast.

**Comment by SNiukalov** [comment]
@LuxoftAKutsan 
 Information inserted from PR description into commits.

**Comment by vkushnirenko-luxoft** [comment]
QNX7_build

**Comment by AByzhynar** [comment]
@SNiukalov Don't forget to add comment to the code

**Review by AByzhynar — APPROVED** [review]

**Review by AKalinich-Luxoft — APPROVED** [review]

**Review by mkorniichuk — APPROVED** [review]

**Comment by LuxoftAKutsan** [comment]
Move it to SmartDevicelinkMirror and close this one
